server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -519,15 +521,28 @@ private static void throwOnCopyToOnObject(Mapper mapper, List<String> copyToFiel
         );
     }
 
-    private static void parseObject(final DocumentParserContext context, String currentFieldName) throws IOException {
+    private static int parseObject(final DocumentParserContext context, String currentFieldName, long arrayObjectsLimit, int alreadyCounted)
+        throws IOException {
+        int newCount = 0;
         assert currentFieldName != null;
         context.setImmediateXContentParent(context.parser().currentToken());
         Mapper objectMapper = context.getMapper(currentFieldName);
         if (objectMapper != null) {
             doParseObject(context, currentFieldName, objectMapper);
         } else {
+            newCount = alreadyCounted + 1;
+            if (newCount > arrayObjectsLimit) {
+                throw new IllegalStateException(
+                    "The number of array objects has exceeded the allowed limit of ["
+                        + arrayObjectsLimit
+                        + "]. This limit can be set by changing the ["
+                        + MapperService.INDEX_MAPPING_ARRAY_OBJECTS_LIMIT_SETTING.getKey()
+                        + "] index level setting."
+                );
+            }
             parseObjectDynamic(context, currentFieldName);
         }
+        return newCount;
```

Comment:
is it necessary to return this ? can't be increment the argument and rely on that?

Reply:
right, sorry, this does not use recursion :) I had to look closer to be reminded. I would leave it as-is.